test/e2e/multicast_test.go

```diff
@@ -158,7 +278,261 @@ type multicastTestcase struct {
+					{PodSelector: &metav1.LabelSelector{MatchLabels: map[string]string{"antrea-e2e": anp.appliedToPod}}},
+				},
+				Egress: []crdv1alpha1.Rule{},
+			},
+		}
+		for i, _ := range anp.ruleConfigs {
+			np.Spec.Egress = append(np.Spec.Egress, crdv1alpha1.Rule{
+				To: []crdv1alpha1.NetworkPolicyPeer{
+					{
+						IPBlock: &crdv1alpha1.IPBlock{
+							CIDR: fmt.Sprintf("%s/32", anp.ruleConfigs[i].address),
+						},
+					},
+				},
+				Ports: []crdv1alpha1.NetworkPolicyPort{
+					{
+						Protocol: &udp,
+					},
+				},
+				Action: &anp.ruleConfigs[i].action,
+				Name:   anp.ruleConfigs[i].name,
+			})
+		}
+		if _, err = k8sUtils.CreateOrUpdateANP(np); err != nil {
+			failOnError(fmt.Errorf("create ANP failed for ANP %s: %v", np.Name, err), t)
+		}
+		failOnError(data.waitForANPRealized(t, data.testNamespace, np.Name), t)
+	}
+
+	for _, anp := range mc.igmpANPConfigs {
+		np := &crdv1alpha1.NetworkPolicy{
+			ObjectMeta: metav1.ObjectMeta{Namespace: data.testNamespace, Name: anp.name, Labels: map[string]string{"antrea-e2e": anp.name}},
+			Spec: crdv1alpha1.NetworkPolicySpec{
+				Priority: p10,
+				AppliedTo: []crdv1alpha1.NetworkPolicyPeer{
+					{PodSelector: &metav1.LabelSelector{MatchLabels: map[string]string{"antrea-e2e": anp.appliedToPod}}},
+				},
+				Egress:  []crdv1alpha1.Rule{},
+				Ingress: []crdv1alpha1.Rule{},
+			},
+		}
+		for i := range anp.ruleConfigs {
+			if anp.ruleConfigs[i].address == "224.0.0.1" {
+				np.Spec.Ingress = append(np.Spec.Ingress, crdv1alpha1.Rule{
+					From: []crdv1alpha1.NetworkPolicyPeer{},
+					Protocols: []crdv1alpha1.NetworkPolicyProtocol{
+						{
+							IGMP: &crdv1alpha1.IGMPProtocol{IGMPType: &igmpQueryType, GroupAddress: "224.0.0.1"},
+						},
+					},
+					Action: &anp.ruleConfigs[i].action,
+					Name:   anp.ruleConfigs[i].name,
+				})
+			} else {
+				np.Spec.Egress = append(np.Spec.Egress, crdv1alpha1.Rule{
+					From: []crdv1alpha1.NetworkPolicyPeer{},
+					Protocols: []crdv1alpha1.NetworkPolicyProtocol{{
+						IGMP: &crdv1alpha1.IGMPProtocol{GroupAddress: anp.ruleConfigs[i].address},
+					}},
+					Action: &anp.ruleConfigs[i].action,
+					Name:   anp.ruleConfigs[i].name,
+				})
+			}
+		}
+		if _, err = k8sUtils.CreateOrUpdateANP(np); err != nil {
+			failOnError(fmt.Errorf("create ANP failed for ANP %s: %v", np.Name, err), t)
+		}
+		failOnError(data.waitForANPRealized(t, data.testNamespace, np.Name), t)
+	}
+
+	for _, receiverConfig := range mc.receiverConfigs {
+		for _, addr := range receiverConfig.IPs {
+			go func(receiver, addr string) {
+				cmd := []string{"/bin/sh", "-c", fmt.Sprintf("mcjoin -o -p %d -W %d %s", 2234, mcjoinWaitTimeout, addr)}
+				data.RunCommandFromPod(data.testNamespace, receiver, mcjoinContainerName, cmd)
+			}(receiverConfig.name, addr)
+		}
+	}
+
+	time.Sleep(5 * time.Second)
+
+	sendSessions := 10
+	var wg sync.WaitGroup
+	for i := 0; i < sendSessions; i++ {
+		wg.Add(1)
+		go func() {
```

Comment:
No need to have this goroutine since every send command is already taken in a single routine in line 435?

Reply:
What is the difference between this solution (10 times * 1 packet) and sending `$session` packets for each senderConfig.IP per sender?

Reply:
updated. The code snippet above has been simplified.